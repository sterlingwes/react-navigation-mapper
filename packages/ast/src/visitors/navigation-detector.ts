import ts from "typescript";

import { AbstractVisitor } from "../abstract-visitor";
import {
  getNavigationDestination,
  getNavigationMethodName,
  importSpecifierIsNativeStackScreenPropsType,
  isValidNavigationMethod,
  propertyAccessIsOnNavigation,
  typeReferenceIsNavigationScreenProp,
} from "./utils/navigation-prop";
import {
  getBlockPosition,
  getVarDeclarationName,
  returnsJSX,
  varDeclarationIsPossibleFunctionComponent,
} from "./utils/component";

type NavigationCall = {
  method: "push" | "replace" | "navigate";
  destination: string;
};

type Component = {
  name: string;
  navigationCalls: NavigationCall[];
};

interface GlobalState {
  components: Array<Component>;
}

interface ModuleState {
  checkProps: boolean;
  insideComponentWithNavigationProp: boolean;
  currentComponent: string | undefined;
  currentComponentPosition: number | undefined;
  currentComponentNavigationCalls: NavigationCall[];
}

export class NavigationDetectorVisitor extends AbstractVisitor<
  GlobalState,
  ModuleState
> {
  constructor() {
    super({
      initialGlobalState: {
        components: [],
      },
      initialModuleState: {
        checkProps: false,
        insideComponentWithNavigationProp: false,
        currentComponent: undefined,
        currentComponentPosition: undefined,
        currentComponentNavigationCalls: [],
      },
    });
  }

  visitor(
    node: ts.Node,
    context: { globalState: GlobalState; moduleState: ModuleState },
    forEachChild: (node: ts.Node) => void
  ) {
    const { moduleState } = context;

    // if (
    //   moduleState.insideComponentWithNavigationProp &&
    //   [ts.SyntaxKind.Identifier].includes(node.kind) === false
    // ) {
    //   console.log(ts.SyntaxKind[node.kind]);
    // }

    const insideComponent = () => {
      return (
        moduleState.currentComponent !== undefined &&
        moduleState.insideComponentWithNavigationProp
      );
    };

    switch (node.kind) {
      case ts.SyntaxKind.VariableDeclaration:
        const varDec = node as ts.VariableDeclaration;
        if (varDeclarationIsPossibleFunctionComponent(varDec)) {
          moduleState.currentComponent = getVarDeclarationName(varDec);
          moduleState.currentComponentPosition = getBlockPosition(varDec);
        }
        break;
      case ts.SyntaxKind.ImportSpecifier:
        const importSpecifier = node as ts.ImportSpecifier;
        if (importSpecifierIsNativeStackScreenPropsType(importSpecifier)) {
          moduleState.checkProps = true;
        }
        break;
      case ts.SyntaxKind.TypeReference:
        // TODO: handle non-object destructuring binding pattern
        if (
          moduleState.checkProps &&
          typeReferenceIsNavigationScreenProp(node as ts.TypeReferenceNode)
        ) {
          moduleState.insideComponentWithNavigationProp = true;
        }
        break;
      // TODO: handle when navigation prop is destructured
      case ts.SyntaxKind.PropertyAccessExpression:
        const propAccess = node as ts.PropertyAccessExpression;
        if (propertyAccessIsOnNavigation(propAccess)) {
          // TODO: capture destination screen and component name reference to establish link
          // using push, replace, navigate methods
          const method = getNavigationMethodName(propAccess);
          if (isValidNavigationMethod(method)) {
            context.moduleState.currentComponentNavigationCalls.push({
              method,
              destination: getNavigationDestination(propAccess),
            });
          }
        }
        break;
    }

    forEachChild(node);

    if (
      node.kind === ts.SyntaxKind.Block &&
      insideComponent() &&
      node.pos === moduleState.currentComponentPosition
    ) {
      context.globalState.components.push({
        name: moduleState.currentComponent!,
        navigationCalls: moduleState.currentComponentNavigationCalls.slice(),
      });
      moduleState.currentComponent = undefined;
      moduleState.currentComponentPosition = undefined;
      moduleState.currentComponentNavigationCalls = [];
      moduleState.insideComponentWithNavigationProp = false;
    }
  }
}
