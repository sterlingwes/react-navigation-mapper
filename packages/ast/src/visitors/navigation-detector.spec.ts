import fs from "fs";
import { describe, it, expect, beforeEach } from "bun:test";
import ts from "typescript";
import { NavigationDetectorVisitor } from "./navigation-detector";

describe("NavigationDetector visitor", () => {
  describe("multi screen simple navigate calls", () => {
    let file: ts.SourceFile;

    beforeEach(() => {
      const fileName =
        "submodules/react-navigation/example/src/Screens/NativeStack.tsx";
      file = ts.createSourceFile(
        fileName,
        fs.readFileSync(fileName).toString(),
        ts.ScriptTarget.ES2015,
        /*setParentNodes */ true
      );
    });

    it("should detect all components & navigation calls", () => {
      const visitor = new NavigationDetectorVisitor();
      visitor.visit(file);

      expect(visitor.globalState.components).toMatchSnapshot();
    });
  });
});
