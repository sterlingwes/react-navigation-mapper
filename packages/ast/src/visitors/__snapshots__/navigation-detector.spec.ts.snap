// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationDetector visitor multi screen simple navigate calls should detect all components & navigation calls 1`] = `
[
  {
    "name": "ArticleScreen",
    "navigationCalls": [
      {
        "destination": "NewsFeed",
        "method": "push",
      },
      {
        "destination": "NewsFeed",
        "method": "replace",
      },
    ],
  },
  {
    "name": "NewsFeedScreen",
    "navigationCalls": [
      {
        "destination": "Albums",
        "method": "push",
      },
    ],
  },
  {
    "name": "AlbumsScreen",
    "navigationCalls": [
      {
        "destination": "Article",
        "method": "navigate",
      },
    ],
  },
]
`;
